core/src/main/java/org/sonarsource/sonarlint/core/AutoBinding.java

```diff
@@ -0,0 +1,408 @@
+import org.sonarsource.sonarlint.core.clientapi.fs.FindFileByNamesInScopeParams;
+import org.sonarsource.sonarlint.core.clientapi.fs.FindFileByNamesInScopeResponse;
+import org.sonarsource.sonarlint.core.clientapi.fs.FoundFile;
+import org.sonarsource.sonarlint.core.commons.log.SonarLintLogger;
+import org.sonarsource.sonarlint.core.commons.progress.ProgressMonitor;
+import org.sonarsource.sonarlint.core.event.BindingConfigChangedEvent;
+import org.sonarsource.sonarlint.core.event.ConfigurationScopeAddedEvent;
+import org.sonarsource.sonarlint.core.event.ConnectionAddedEvent;
+import org.sonarsource.sonarlint.core.serverapi.ServerApi;
+import org.sonarsource.sonarlint.core.serverapi.component.ServerProject;
+
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+import static org.apache.commons.lang.StringUtils.removeEnd;
+import static org.apache.commons.lang.StringUtils.trimToNull;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+public class AutoBinding {
+
+  private static final SonarLintLogger LOG = SonarLintLogger.get();
+  public static final String SONAR_SCANNER_CONFIG_FILENAME = "sonar-project.properties";
+  public static final String AUTOSCAN_CONFIG_FILENAME = ".sonarcloud.properties";
+  private static final String SONARCLOUD_URL = "https://sonarcloud.io";
+
+  private final ConfigurationReferential configReferential;
+  private final ConnectionConfigurationReferential connectionReferential;
+  private final SonarLintClient client;
+
+  public AutoBinding(ConfigurationReferential configReferential, ConnectionConfigurationReferential connectionReferential, SonarLintClient client) {
+    this.configReferential = configReferential;
+    this.connectionReferential = connectionReferential;
+    this.client = client;
+  }
+
+  @Subscribe
+  public void bindingConfigChanged(BindingConfigChangedEvent event) {
+    // Check if auto-bind is switched on
+    if (event.getNewConfig().isAutoBindEnabled() && !event.getPreviousConfig().isAutoBindEnabled()) {
+      autoBindConfigScope(event.getNewConfig().getConfigScopeId());
+    }
+  }
+
+  @Subscribe
+  public void configurationScopeAdded(ConfigurationScopeAddedEvent event) {
+    var configScopeId = event.getAddedConfigurationScopeId();
+    var bindingConfiguration = configReferential.getBindingConfiguration(configScopeId);
+    if (bindingConfiguration == null) {
+      // Maybe the configuration was removed since the event was raised
+      LOG.debug("Configuration scope '{}' not found. Ignoring event.", configScopeId);
+      return;
+    }
+    if (bindingConfiguration.isAutoBindEnabled()) {
+      autoBindConfigScope(configScopeId);
+    }
+  }
+
+  @Subscribe
+  public void connectionAdded(ConnectionAddedEvent event) {
+    // Double check if added connection has not been removed in the meantime
+    if (connectionReferential.getConnectionById(event.getAddedConnectionId()) != null) {
+      autoBindAll();
```

Comment:
First, I was a bit surprised that the just added connection is not used at all to filter binding suggestions, which means we can suggest the user to bind to a different connection. I wonder if this wouldn't be too noisy, as the user would have probably already received a notification for that (by e.g. opening a project)

Reply:
I don't think the current behavior is wrong (if the user previously received a notification, he might have marked the project as "don't auto-bind").
I will see how to do better in another PR.